From 08a6472126438a9f6ef233d3e644543c025171a9 Mon Sep 17 00:00:00 2001
From: DrCookie <timothee.vanhove@heig-vd.ch>
Date: Sun, 3 Dec 2023 18:05:09 +0100
Subject: [PATCH] small refactor do_renice

---
 so3/kernel/process.c | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/so3/kernel/process.c b/so3/kernel/process.c
index d736ac7..7baee0d 100644
--- a/so3/kernel/process.c
+++ b/so3/kernel/process.c
@@ -98,14 +98,12 @@ pcb_t *find_proc_by_pid(uint32_t pid) {
 }
 
 int do_renice(uint32_t pid, uint32_t new_prio) {
-	//TO COMPLETE
 	if(pid < 0 || new_prio < 1) {
 		return -1;
-	} else {
-		pcb_t* pcb = find_proc_by_pid(pid);
-		pcb->main_thread->prio = new_prio;
 	}
-
+    
+	pcb_t* pcb = find_proc_by_pid(pid);
+	pcb->main_thread->prio = new_prio;
 	return 0;
 }
 
-- 
GitLab

